pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ */
+package io.pravega.example.streamprocessing;
+
+import io.pravega.client.ClientFactory;
+import io.pravega.client.admin.ReaderGroupManager;
+import io.pravega.client.admin.StreamManager;
+import io.pravega.client.stream.*;
+import io.pravega.client.stream.impl.UTF8StringSerializer;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * A simple example that demonstrates reading events from a Pravega stream, processing each event,
+ * and writing each output event to another Pravega stream.
+ *
+ * This supports multiple worker threads.
+ * Upon restart, it restarts from the last successful checkpoint and guarantees exactly-once semantics.
+ * Workers must be stateless.
+ *
+ * Use {@link EventGenerator} to generate input events and {@link EventDebugSink}
+ * to view the output events.
+ *
+ * See also {@link ExactlyOnceMultithreadedProcessorWorker}.
+ */
+
+public class ExactlyOnceMultithreadedProcessor implements Callable<Void> {
```

Comment:
It would be good to add a bunch of comments here, at the top of the file and throughout explaining the overall design, what it's doing and why.